## 3.1. Intro to Debugging
---
** Learning goals**
- strategi untuk menghindari errors
- kesalahan umum dan penyebabnya

** Tujuan **
- identifikasi Syntax errors berdasarkan errors messages
- menemukan ValueError, TypeError, SytaxError, ParseError, dan NameError


## 3.2. Debugging
---
dalam proses pemograman sering terjadi kesalahan atau bug. sedangkan proses untuk melacak dan memperbaikinya disebut dengan debugging.
 
untuk menghindari kesalahan aau error, berikut hal-hal yang dapat dilakukan,
- memahami masalah serta tujuan yang ingin dicapai sehingga dapat mengidentifikasi solusi yang benar
- start small
- progress and keep improving

---
Ask the user for the time now (in hours 0 – 23), and ask for the number of hours to wait. Your program should output what the time will be on the clock when the alarm goes off.

So, where to start? The problem requires two pieces of input from the user, so lets start there and make sure we can get the data we need.

In [1]:
current_time = input("what is the current time (in hours)?")
wait_time = input("How many hours do you want to wait")

print(current_time)
print(wait_time)

10
2


In [2]:
#debugging 1
current_time = input("what is the current time (in hours 0--23)?")
wait_time = input("How many hours do you want to wait")

print(current_time)
print(wait_time)

final_time = current_time + wait_time
print(final_time)

10
2
102


menghasilkan 102, karena pada input 10 dan 2 dibaca sebagai string bukan integer

In [3]:
#debugging 2
current_time_str = input("what is the current time (in hours 0-23)?")
wait_time_str = input("How many hours do you want to wait")

current_time_int = int(current_time_str)
wait_time_int = int(wait_time_str)

final_time_int = current_time_int + wait_time_int
print(final_time_int)

12


ketika menggunakan waktu yang lebih besar maka akan terbaca jam yang lebih dari 24, maka dibutuhkan pembagian bersisa sehingga waktu yang dihasilkan tetap berada pada interval waktu (24)

In [4]:
#debugging 3
current_time_str = input("what is the current time (in hours 0-23)?")
wait_time_str = input("How many hours do you want to wait")

current_time_int = int(current_time_str)
wait_time_int = int(wait_time_str)

final_time_int = current_time_int + wait_time_int

final_answer = final_time_int % 24

print("The time after waiting is: ", final_answer)

The time after waiting is:  12


## 3.3. Beginning Tips
---
1. kesalahan dapat terjadi pada segala aspek termasuk Python sendiri, namun kesalahan pada Python harus menjadi aspek kesalahan terakhir ketika muncul error
2. pengujian asumsi pada setiap kemungkinan
3. menggunakan petunjuk seperti Error messages dan print statment

## 3.4. Syntax Error
---
Python dapat menjalakan suatu program ketika secara sintaks terpenuhi. ketika tidak terpenuhi maka gagal dan muncul error message. sintaks sendiri merupakan struktur dari program dan aturannya.

dalam Python sendiri, 1 kesalahan dapat langsung menyebabkan sintaks tidak terpenuhi dan proses terhenti.

In [5]:
print("Hello World!"

SyntaxError: unexpected EOF while parsing (<ipython-input-5-0219726117bb>, line 1)

kesalahan: ) diakhir

In [6]:
print("Hello World!")

Hello World!


## 3.5. Runtime Errors
---
selain SyntaxError, kesalahan lain adalah Runtime Errors yaitu kesalahan yang muncul setelah program dijalankan. kesalahan ini jarang muncul pada program sederhana.

In [7]:
A = int(input("angka pertama"))
B = int(input("angka kedua"))

pembagian = A/B

print("hasil pembagian angka pertama dan kedua", pembagian)

ZeroDivisionError: division by zero

terjadi kesalahan karena input angka kedua adalah 0 atau menghasilkan inf

## 3.6. Semantic Errors
---
merupakan kesalahan yang terjadi dimana program akan berjalan namun akan menghasilkan output yang salah atau tidak sesuai dengan perintah tanpa adanya error message.

jenis kesalahan ini lebih sulit diuraikan karena diperlukan pemahaman dan ketelitian

In [9]:
angka = int(input("angka yang akan dibuat persentase"))

persen = angka

print("hasil persentase",persen,"%")

hasil persentase 12 %


terjadi kesalahan hasil dari persentase 12 adalah 0.12 %

In [10]:
angka = int(input("angka yang akan dibuat persentase"))

persen = angka/100

print("hasil persentase",persen,"%")

hasil persentase 0.12 %
